src/renderer/components/swap/Swap.tsx

```diff
@@ -390,6 +389,156 @@ export const Swap = ({
     )
   }, [assetsToSwap, onConfirmSwap, changeAmount, closePrivateModal])
 
+  const sourceChainFee: RD.RemoteData<Error, BaseAmount> = useMemo(
+    () =>
+      FP.pipe(
+        feesProp,
+        RD.map(({ source }) => source)
+      ),
+    [feesProp]
+  )
+
+  const targetChainFee: RD.RemoteData<Error, BaseAmount> = useMemo(
+    () =>
+      FP.pipe(
+        feesProp,
+        RD.map(({ target }) => target)
+      ),
+    [feesProp]
+  )
+
+  const sourceChainBalanceError: boolean = useMemo(
+    () =>
+      FP.pipe(
+        sequenceTOption(sourceAsset, walletBalances),
+        O.map(([sourceAsset, walletBalances]) => {
+          // If sourceAsset is the same as source chain
+          // we should subtract changeAmount
+          const shouldSubtractChangeAmount = eqAsset.equals(sourceAsset, getChainAsset(sourceAsset.chain))
+
+          const balanceMinusAmount = FP.pipe(
+            walletBalances,
+            A.findFirst((walletBalance) => eqAsset.equals(walletBalance.asset, sourceAsset)),
+            O.map((sourceBalance) =>
+              baseToAsset(sourceBalance.amount)
+                .amount()
+                .minus(shouldSubtractChangeAmount ? changeAmount : ZERO_BN)
+            ),
+            O.map(FP.flow(assetAmount, assetToBase))
+          )
+
+          return FP.pipe(
+            sequenceTOption(balanceMinusAmount, FP.pipe(sourceChainFee, RD.toOption)),
+            O.map(([leftBalance, sourceChainFee]) => {
+              const sourceChainFeeAmount = sourceChainFee.amount()
+
+              return leftBalance.amount().minus(sourceChainFeeAmount)
+            }),
+            O.map((result) => result.isNegative()),
+            O.getOrElse((): boolean => false)
+          )
+        }),
+        O.getOrElse((): boolean => false)
+      ),
+    [changeAmount, sourceAsset, walletBalances, sourceChainFee]
+  )
+
+  const sourceChainErrorLabel: JSX.Element = useMemo(() => {
+    if (!sourceChainBalanceError) {
+      return <></>
+    }
+    const sourceChainAsset = FP.pipe(
+      sourceAsset,
+      O.map(({ chain }) => getChainAsset(chain)),
+      O.map((asset) => asset.symbol),
+      O.getOrElse(() => '')
+    )
+
+    return (
+      <Styled.BalanceErrorLabel>
+        {intl.formatMessage({ id: 'swap.errors.amount.balanceShouldCoverChainFee' }, { asset: sourceChainAsset })}{' '}
+      </Styled.BalanceErrorLabel>
+    )
+  }, [sourceChainBalanceError, sourceAsset, intl])
+
+  const targetChainFeeAmountInTargetAsset: BaseAmount = useMemo(() => {
+    const chainFee = FP.pipe(
+      targetChainFee,
+      RD.getOrElse(() => ZERO_BASE_AMOUNT)
+    )
+
+    return FP.pipe(
+      targetAsset,
+      O.map((asset) => {
+        const chainAsset = getChainAsset(asset.chain)
+        return getValueOfAsset1InAsset2(chainFee, poolData[assetToString(chainAsset)], poolData[assetToString(asset)])
```

Comment:
yep. my bad. should check for `undefined`
but im gonna to disagree with using `getPoolDetail`. it uses `Array.prototype.find` method. it means we we need go through ALL items everytime we want to use it. this is why here was used a hash map

Reply:
> it uses Array.prototype.find method.

It's too early to worry about such performance things ... It can be done in 2022.

Reply:
> It's too early to worry about such performance things ... 

this is really dangerous question 😄 when is not it 'early' ? when we will see it by our eyes ? it will be too late IMO 😄 

hm? 
> will avoid to write extra code

`getPoolDetail` will return `O.Option`.  we will need to combine 2 options here and after use `toPoolData`.
seems this is the same as compare 2 values with `undefined` here but without handling each element in an array